File: src/scripts/download_sentencias.js
/**
 * Descarga de sentencias de la Sala 4 del Tribunal Supremo desde CENDOJ.
 *
 * Recorre las páginas de resultados de búsqueda, extrae metadatos de cada sentencia
 * y descarga los PDFs correspondientes.
 *
 * Uso:
 *   node download_sentencias.js                         # Descarga todas (hasta 200)
 *   node download_sentencias.js --max-pages 1           # Solo primera página
 *   node download_sentencias.js --skip-pdf              # Solo metadatos
 *   node download_sentencias.js --sala-social           # Filtrar solo Sala de lo Social
 *   node download_sentencias.js --resume                # Reanudar descarga
 *
 * No requiere dependencias externas (usa fetch nativo de Node 18+).
 */

const fs = require('fs');
const path = require('path');
const { URL } = require('url');

// --- Configuración -----------------------------------------------------------

const BASE_URL = 'https://www.poderjudicial.es';
const DEFAULT_QUERY = 'sistema de la seguridad social';
const MAX_PAGES_DEFAULT = 20;
const RESULTS_PER_PAGE = 10;
const DELAY_MS = 1500;
const MAX_RETRIES = 3;

const PROJECT_ROOT = path.resolve(__dirname, '..', '..');
const OUTPUT_DIR = path.join(PROJECT_ROOT, 'data', 'sentencias');
const PDF_DIR = path.join(OUTPUT_DIR, 'pdf');
const METADATA_FILE = path.join(OUTPUT_DIR, 'sentencias_metadata.json');
const PROGRESS_FILE = path.join(OUTPUT_DIR, 'download_progress.json');

const HEADERS = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
    'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
    'Accept-Language': 'es-ES,es;q=0.9,en;q=0.8',
};

// --- Argumentos CLI ----------------------------------------------------------

function parseArgs() {
    const args = {
        query: DEFAULT_QUERY,
        maxPages: MAX_PAGES_DEFAULT,
        delay: DELAY_MS,
        skipPdf: false,
        salaSocial: false,
        resume: false,
    };
    const argv = process.argv.slice(2);
    for (let i = 0; i < argv.length; i++) {
        switch (argv[i]) {
            case '--query': case '-q': args.query = argv[++i]; break;
            case '--max-pages': case '-p': args.maxPages = parseInt(argv[++i]); break;
            case '--delay': case '-d': args.delay = parseFloat(argv[++i]) * 1000; break;
            case '--skip-pdf': args.skipPdf = true; break;
            case '--sala-social': args.salaSocial = true; break;
            case '--resume': args.resume = true; break;
            case '--help': case '-h':
                console.log(`
Descarga sentencias del TS (Sala Social) desde CENDOJ

Opciones:
  --query, -q <texto>     Termino de busqueda (default: "${DEFAULT_QUERY}")
  --max-pages, -p <n>     Maximo de paginas (default: ${MAX_PAGES_DEFAULT})
  --delay, -d <seg>       Segundos entre peticiones (default: ${DELAY_MS/1000})
  --skip-pdf              Solo extraer metadatos, no descargar PDFs
  --sala-social           Filtrar solo Sala de lo Social
  --resume                Reanudar descarga anterior
  --help, -h              Mostrar esta ayuda
`);
                process.exit(0);
        }
    }
    return args;
}

// --- Utilidades --------------------------------------------------------------

function sleep(ms) { return new Promise(r => setTimeout(r, ms)); }

function getOffset(pageNum) {
    // Página 1 → /1/, página 2 → /11/, página 3 → /21/
    return pageNum === 1 ? 1 : (pageNum - 1) * 10 + 1;
}

function sanitizeFilename(roj) {
    return roj.replace(/[^\w]/g, '_').replace(/^_+|_+$/g, '');
}

function ensureDir(dir) {
    if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
}

function loadJson(filepath, defaultVal) {
    try {
        return JSON.parse(fs.readFileSync(filepath, 'utf-8'));
    } catch { return typeof defaultVal === 'function' ? defaultVal() : defaultVal; }
}

function saveJson(filepath, data) {
    fs.writeFileSync(filepath, JSON.stringify(data, null, 2), 'utf-8');
}

// --- Fetch con reintentos ----------------------------------------------------

async function fetchPage(url) {
    for (let attempt = 1; attempt <= MAX_RETRIES; attempt++) {
        try {
            const resp = await fetch(url, { headers: HEADERS, signal: AbortSignal.timeout(30000) });
            if (!resp.ok) throw new Error(`HTTP ${resp.status} ${resp.statusText}`);
            return await resp.text();
        } catch (e) {
            console.log(`  [!] Intento ${attempt}/${MAX_RETRIES} fallido: ${e.message}`);
            if (attempt < MAX_RETRIES) {
                const wait = Math.pow(2, attempt) * 1000;
                console.log(`  Esperando ${wait/1000}s...`);
                await sleep(wait);
            }
        }
    }
    return null;
}

// --- Parser HTML simple (sin dependencias) -----------------------------------

/**
 * Parsea el HTML de resultados CENDOJ sin BeautifulSoup.
 * Busca los links a /search/documento/ y extrae los metadatos del bloque circundante.
 */
function parseSearchResults(html, query) {
    const sentencias = [];

    // Regex para encontrar links a documentos con ROJ/ECLI
    const linkRegex = /<a[^>]*href="([^"]*\/search\/documento\/[^"]+)"[^>]*>([\s\S]*?)<\/a>/gi;
    let linkMatch;

    while ((linkMatch = linkRegex.exec(html)) !== null) {
        const href = linkMatch[1];
        const linkText = linkMatch[2].replace(/<[^>]+>/g, '').trim();

        // Extraer ROJ y ECLI del texto del link
        const rojMatch = linkText.match(/ROJ:\s*(.+?)\s*-\s*(ECLI:\S+)/);
        if (!rojMatch) continue;

        const roj = rojMatch[1].trim();
        const ecli = rojMatch[2].trim();

        // Extraer reference_id y optimize_date de la URL
        const hrefMatch = href.match(/\/search\/documento\/\w+\/(\d+)\/.*?\/(\d+)/);
        if (!hrefMatch) continue;

        const referenceId = hrefMatch[1];
        const optimizeDate = hrefMatch[2];

        // Extraer bloque de texto alrededor del link para obtener metadatos
        // Buscamos desde el link hasta el siguiente link o fin de contenido
        const linkPos = linkMatch.index;
        const nextLinkPos = html.indexOf('/search/documento/', linkPos + linkMatch[0].length);
        const blockEnd = nextLinkPos > 0 ? nextLinkPos : linkPos + 3000;
        const block = html.substring(linkPos, Math.min(blockEnd, linkPos + 3000));

        // Limpiar HTML tags del bloque
        const blockText = block.replace(/<[^>]+>/g, '\n').replace(/&nbsp;/g, ' ');

        function extractField(pattern, defaultVal = '') {
            const m = blockText.match(pattern);
            return m ? m[1].trim().replace(/-\s*$/, '').trim() : defaultVal;
        }

        const sentencia = {
            roj,
            ecli,
            reference_id: referenceId,
            optimize_date: optimizeDate,
            tipo_organo: extractField(/Tipo\s+.rgano:\s*(.+?)(?:\n|Municipio:)/i),
            municipio: extractField(/Municipio:\s*(.+?)(?:\n|Ponente:)/i),
            ponente: extractField(/Ponente:\s*(.+?)(?:\n|N.\s*Recurso:)/i),
            recurso: extractField(/N.\s*Recurso:\s*(.+?)(?:\n|Fecha:)/i),
            fecha: extractField(/Fecha:\s*(.+?)(?:\n|Tipo\s+Resoluci)/i),
            tipo_resolucion: extractField(/Tipo\s+Resoluci.n:\s*(.+?)(?:\n|Resumen:)/i),
            resumen: extractField(/Resumen:\s*([\s\S]+?)(?:\n\s*\n|Icono compartir|compartir en)/i),
            url_documento: href.startsWith('/') ? BASE_URL + href : href,
            url_pdf: `${BASE_URL}/search/contenidos.action?action=accessToPDF&publicinterface=true&tab=AN&reference=${referenceId}&optimize=${optimizeDate}&links=${encodeURIComponent(query)}&databasematch=TS`,
            pdf_filename: sanitizeFilename(roj) + '.pdf',
        };

        // Limpiar resumen
        sentencia.resumen = sentencia.resumen.replace(/\s+/g, ' ').trim();

        sentencias.push(sentencia);
    }

    return sentencias;
}

// --- Descarga de PDFs --------------------------------------------------------

async function downloadPdf(sentencia) {
    const pdfPath = path.join(PDF_DIR, sentencia.pdf_filename);

    // Saltar si ya existe
    try {
        const stat = fs.statSync(pdfPath);
        if (stat.size > 0) return true;
    } catch {}

    for (let attempt = 1; attempt <= MAX_RETRIES; attempt++) {
        try {
            const resp = await fetch(sentencia.url_pdf, {
                headers: HEADERS,
                signal: AbortSignal.timeout(60000),
            });
            if (!resp.ok) throw new Error(`HTTP ${resp.status}`);

            const contentType = resp.headers.get('content-type') || '';
            if (contentType.includes('html')) {
                console.log('\n  [!] Respuesta HTML en vez de PDF');
                return false;
            }

            const buffer = Buffer.from(await resp.arrayBuffer());
            if (buffer.length < 1024) {
                console.log(`\n  [!] PDF sospechosamente pequeno (${buffer.length} bytes)`);
                return false;
            }

            fs.writeFileSync(pdfPath, buffer);
            return true;

        } catch (e) {
            console.log(`\n  [!] Intento ${attempt}/${MAX_RETRIES} fallido: ${e.message}`);
            if (attempt < MAX_RETRIES) await sleep(Math.pow(2, attempt) * 1000);
        }
    }
    return false;
}

// --- Main --------------------------------------------------------------------

async function main() {
    const args = parseArgs();
    const queryEncoded = encodeURIComponent(args.query);

    ensureDir(OUTPUT_DIR);
    ensureDir(PDF_DIR);

    let progress = loadJson(PROGRESS_FILE, () => ({ pages_scraped: [], pdfs_downloaded: [], pdfs_failed: [] }));
    let existingMetadata = loadJson(METADATA_FILE, []);
    const existingRoj = new Set(existingMetadata.map(s => s.roj));

    console.log('='.repeat(70));
    console.log('DESCARGA DE SENTENCIAS - CENDOJ (Poder Judicial)');
    console.log('='.repeat(70));
    console.log(`  Busqueda: "${args.query}"`);
    console.log(`  Paginas:  hasta ${args.maxPages} (max. ${args.maxPages * RESULTS_PER_PAGE} sentencias)`);
    console.log(`  Delay:    ${args.delay/1000}s entre peticiones`);
    console.log(`  Salida:   ${OUTPUT_DIR}`);
    if (args.salaSocial) console.log(`  Filtro:   Solo Sala de lo Social`);
    if (args.skipPdf) console.log(`  PDFs:     NO (solo metadatos)`);
    if (args.resume && existingMetadata.length) console.log(`  Reanudando: ${existingMetadata.length} sentencias previas`);
    console.log('='.repeat(70));
    console.log();

    let allSentencias = args.resume ? [...existingMetadata] : [];
    let allRoj = args.resume ? new Set(existingRoj) : new Set();
    let newCount = 0;
    let skippedCount = 0;

    // == Fase 1: Scraping de metadatos ========================================

    console.log('FASE 1: Extraccion de metadatos de resultados de busqueda');
    console.log('-'.repeat(50));

    for (let page = 1; page <= args.maxPages; page++) {
        if (args.resume && progress.pages_scraped.includes(page)) {
            console.log(`  Pagina ${String(page).padStart(2)}/${args.maxPages}: ya procesada, saltando`);
            continue;
        }

        const offset = getOffset(page);
        const url = `${BASE_URL}/search/sentencias/${queryEncoded}/${offset}/PUB`;
        console.log(`  Pagina ${String(page).padStart(2)}/${args.maxPages}: descargando...`);

        const html = await fetchPage(url);
        if (!html) {
            console.log(`  [X] No se pudo obtener la pagina ${page}`);
            continue;
        }

        let sentencias = parseSearchResults(html, args.query);

        if (sentencias.length === 0) {
            console.log(`  [X] Sin resultados en pagina ${page} -- fin de resultados`);
            break;
        }

        if (args.salaSocial) {
            const before = sentencias.length;
            sentencias = sentencias.filter(s =>
                (s.tipo_organo || '').toLowerCase().includes('social')
            );
            if (before !== sentencias.length) {
                console.log(`    (filtradas ${before - sentencias.length} no-Social)`);
            }
        }

        let pageNew = 0;
        for (const s of sentencias) {
            if (!allRoj.has(s.roj)) {
                allSentencias.push(s);
                allRoj.add(s.roj);
                pageNew++;
            } else {
                skippedCount++;
            }
        }

        newCount += pageNew;
        console.log(`    -> ${sentencias.length} sentencias, ${pageNew} nuevas`);

        progress.pages_scraped.push(page);
        saveJson(METADATA_FILE, allSentencias);
        saveJson(PROGRESS_FILE, progress);

        if (page < args.maxPages) await sleep(args.delay);
    }

    console.log();
    console.log(`Total sentencias recopiladas: ${allSentencias.length} (${newCount} nuevas)`);
    if (skippedCount) console.log(`  (${skippedCount} duplicadas omitidas)`);
    saveJson(METADATA_FILE, allSentencias);

    if (args.skipPdf) {
        console.log();
        console.log(`[OK] Metadatos guardados en: ${METADATA_FILE}`);
        return;
    }

    // == Fase 2: Descarga de PDFs =============================================

    console.log();
    console.log('FASE 2: Descarga de PDFs');
    console.log('-'.repeat(50));

    const downloadedSet = new Set(progress.pdfs_downloaded || []);
    const failedSet = new Set(progress.pdfs_failed || []);

    const toDownload = allSentencias.filter(s =>
        !downloadedSet.has(s.roj) && !failedSet.has(s.roj)
    );

    let already = 0;
    for (const s of allSentencias) {
        try {
            const stat = fs.statSync(path.join(PDF_DIR, s.pdf_filename));
            if (stat.size > 0) already++;
        } catch {}
    }

    console.log(`  ${already} ya descargados, ${toDownload.length} pendientes`);
    console.log();

    let success = already;
    let failed = 0;

    for (let i = 0; i < toDownload.length; i++) {
        const sentencia = toDownload[i];
        const roj = sentencia.roj;
        const pdfPath = path.join(PDF_DIR, sentencia.pdf_filename);

        process.stdout.write(`  [${i+1}/${toDownload.length}] ${roj} (${sentencia.fecha || '?'})... `);

        // Comprobar si ya existe
        try {
            const stat = fs.statSync(pdfPath);
            if (stat.size > 0) {
                console.log('ya existe [OK]');
                progress.pdfs_downloaded.push(roj);
                success++;
                saveJson(PROGRESS_FILE, progress);
                continue;
            }
        } catch {}

        const ok = await downloadPdf(sentencia);
        if (ok) {
            const stat = fs.statSync(pdfPath);
            const sizeKb = Math.round(stat.size / 1024);
            console.log(`[OK] (${sizeKb} KB)`);
            progress.pdfs_downloaded.push(roj);
            success++;
        } else {
            console.log('[FALLO]');
            progress.pdfs_failed.push(roj);
            failed++;
        }

        saveJson(PROGRESS_FILE, progress);
        await sleep(args.delay);
    }

    // == Resumen ==============================================================

    console.log();
    console.log('='.repeat(70));
    console.log('RESUMEN');
    console.log('='.repeat(70));
    console.log(`  Sentencias totales:   ${allSentencias.length}`);
    console.log(`  PDFs descargados:     ${success}`);
    console.log(`  PDFs fallidos:        ${failed}`);
    console.log(`  Metadatos:            ${METADATA_FILE}`);
    console.log(`  PDFs:                 ${PDF_DIR}`);

    if (failed > 0) {
        console.log();
        console.log('  Para reintentar los fallidos, ejecuta de nuevo con --resume');
    }
    console.log();
}

main().catch(e => {
    console.error('Error fatal:', e);
    process.exit(1);
});

File: src/scripts/download_sentencias.py
"""
Descarga de sentencias de la Sala 4 del Tribunal Supremo desde CENDOJ.

Recorre las páginas de resultados de búsqueda, extrae metadatos de cada sentencia
y descarga los PDFs correspondientes.

Uso:
    python download_sentencias.py                    # Descarga todas (hasta 200)
    python download_sentencias.py --max-pages 1      # Solo primera página (10 sentencias)
    python download_sentencias.py --query "despido"  # Cambiar término de búsqueda
    python download_sentencias.py --sala-social      # Filtrar solo Sala de lo Social

Requisitos:
    pip install requests beautifulsoup4
"""

import argparse
import json
import re
import sys
import time
from pathlib import Path
from urllib.parse import quote

try:
    import requests
    from bs4 import BeautifulSoup
except ImportError:
    print("ERROR: Faltan dependencias. Ejecuta:")
    print("  pip install requests beautifulsoup4")
    sys.exit(1)

# --- Configuracion por defecto ------------------------------------------------

BASE_URL = "https://www.poderjudicial.es"
SEARCH_URL = BASE_URL + "/search/sentencias/{query}/{offset}/PUB"
PDF_URL = (
    BASE_URL + "/search/contenidos.action"
    "?action=accessToPDF&publicinterface=true&tab=AN"
    "&reference={reference}&optimize={optimize}"
    "&links={query}&databasematch=TS"
)

DEFAULT_QUERY = "sistema de la seguridad social"
MAX_PAGES = 20          # CENDOJ limita a 200 resultados (20 paginas x 10)
RESULTS_PER_PAGE = 10
DELAY_BETWEEN_REQUESTS = 1.5  # segundos entre peticiones
MAX_RETRIES = 3

# Rutas de salida
PROJECT_ROOT = Path(__file__).resolve().parent.parent.parent
OUTPUT_DIR = PROJECT_ROOT / "data" / "sentencias"
PDF_DIR = OUTPUT_DIR / "pdf"
METADATA_FILE = OUTPUT_DIR / "sentencias_metadata.json"
PROGRESS_FILE = OUTPUT_DIR / "download_progress.json"

# Headers para simular navegador
HEADERS = {
    "User-Agent": (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
        "AppleWebKit/537.36 (KHTML, like Gecko) "
        "Chrome/120.0.0.0 Safari/537.36"
    ),
    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
    "Accept-Language": "es-ES,es;q=0.9,en;q=0.8",
}


# --- Funciones auxiliares -----------------------------------------------------

def get_offset(page_num):
    """Calcula el offset de URL para una pagina dada (1-indexed).
    Patron CENDOJ: pagina 1 -> /1/, pagina 2 -> /11/, pagina 3 -> /21/, etc.
    """
    if page_num == 1:
        return 1
    return (page_num - 1) * 10 + 1


def sanitize_filename(roj):
    """Convierte ROJ en nombre de archivo valido. 'STS 6045/2025' -> 'STS_6045_2025'"""
    return re.sub(r'[^\w]', '_', roj).strip('_')


def load_progress():
    """Carga el progreso de descargas anteriores."""
    if PROGRESS_FILE.exists():
        with open(PROGRESS_FILE, 'r', encoding='utf-8') as f:
            return json.load(f)
    return {"pages_scraped": [], "pdfs_downloaded": [], "pdfs_failed": []}


def save_progress(progress):
    """Guarda el progreso actual."""
    with open(PROGRESS_FILE, 'w', encoding='utf-8') as f:
        json.dump(progress, f, indent=2, ensure_ascii=False)


def load_metadata():
    """Carga metadatos guardados previamente."""
    if METADATA_FILE.exists():
        with open(METADATA_FILE, 'r', encoding='utf-8') as f:
            return json.load(f)
    return []


def save_metadata(metadata):
    """Guarda los metadatos."""
    with open(METADATA_FILE, 'w', encoding='utf-8') as f:
        json.dump(metadata, f, indent=2, ensure_ascii=False)


def fetch_page(url, session):
    """Descarga una pagina con reintentos."""
    for attempt in range(1, MAX_RETRIES + 1):
        try:
            resp = session.get(url, headers=HEADERS, timeout=30)
            resp.raise_for_status()
            return resp.text
        except requests.RequestException as e:
            print("  [!] Intento {}/{} fallido: {}".format(attempt, MAX_RETRIES, e))
            if attempt < MAX_RETRIES:
                wait = 2 ** attempt
                print("  Esperando {}s antes de reintentar...".format(wait))
                time.sleep(wait)
    return None


# --- Scraping de resultados ---------------------------------------------------

def parse_search_results(html, query):
    """Extrae metadatos de sentencias de una pagina de resultados."""
    soup = BeautifulSoup(html, 'html.parser')
    sentencias = []

    # Encontrar todos los links a documentos individuales
    doc_links = soup.find_all('a', href=re.compile(r'/search/documento/'))

    for link in doc_links:
        href = link.get('href', '')
        text = link.get_text(strip=True)

        # Parsear ROJ y ECLI del texto del link
        roj_match = re.search(r'ROJ:\s*(.+?)\s*-\s*(ECLI:\S+)', text)
        if not roj_match:
            continue

        roj = roj_match.group(1).strip()
        ecli = roj_match.group(2).strip()

        # Extraer reference_id y optimize_date del href
        href_match = re.search(r'/search/documento/\w+/(\d+)/.*?/(\d+)', href)
        if not href_match:
            continue

        reference_id = href_match.group(1)
        optimize_date = href_match.group(2)

        # Navegar al contenedor padre para extraer metadatos
        container = link
        for _ in range(10):
            parent = container.parent
            if parent is None:
                break
            parent_text = parent.get_text()
            if 'Tipo' in parent_text and 'Resumen' in parent_text:
                container = parent
                break
            container = parent

        block_text = container.get_text(separator='\n')

        def extract_field(pattern, txt, default=""):
            m = re.search(pattern, txt)
            return m.group(1).strip().rstrip('-').strip() if m else default

        sentencia = {
            "roj": roj,
            "ecli": ecli,
            "reference_id": reference_id,
            "optimize_date": optimize_date,
            "tipo_organo": extract_field(
                r'Tipo .rgano:\s*(.+?)(?:\n|Municipio:)', block_text
            ),
            "municipio": extract_field(
                r'Municipio:\s*(.+?)(?:\n|Ponente:)', block_text
            ),
            "ponente": extract_field(
                r'Ponente:\s*(.+?)(?:\n|N. Recurso:)', block_text
            ),
            "recurso": extract_field(
                r'N. Recurso:\s*(.+?)(?:\n|Fecha:)', block_text
            ),
            "fecha": extract_field(
                r'Fecha:\s*(.+?)(?:\n|Tipo Resoluci)', block_text
            ),
            "tipo_resolucion": extract_field(
                r'Tipo Resoluci.n:\s*(.+?)(?:\n|Resumen:)', block_text
            ),
            "resumen": extract_field(
                r'Resumen:\s*(.+?)(?:\n\s*\n|Icono compartir|$)', block_text
            ),
            "url_documento": BASE_URL + href if href.startswith('/') else href,
            "url_pdf": PDF_URL.format(
                reference=reference_id,
                optimize=optimize_date,
                query=quote(query, safe='')
            ),
            "pdf_filename": sanitize_filename(roj) + ".pdf",
        }

        # Limpiar resumen multilinea
        sentencia["resumen"] = re.sub(r'\s+', ' ', sentencia["resumen"]).strip()

        sentencias.append(sentencia)

    return sentencias


# --- Descarga de PDFs ---------------------------------------------------------

def download_pdf(sentencia, session):
    """Descarga el PDF de una sentencia. Retorna True si se descargo correctamente."""
    pdf_path = PDF_DIR / sentencia["pdf_filename"]

    if pdf_path.exists() and pdf_path.stat().st_size > 0:
        return True

    url = sentencia["url_pdf"]

    for attempt in range(1, MAX_RETRIES + 1):
        try:
            resp = session.get(url, headers=HEADERS, timeout=60, stream=True)
            resp.raise_for_status()

            content_type = resp.headers.get('Content-Type', '')
            if 'html' in content_type.lower():
                print("\n  [!] Respuesta HTML en vez de PDF")
                return False

            with open(pdf_path, 'wb') as f:
                for chunk in resp.iter_content(chunk_size=8192):
                    f.write(chunk)

            size_kb = pdf_path.stat().st_size / 1024
            if size_kb < 1:
                print("\n  [!] PDF sospechosamente pequeno ({:.1f} KB)".format(size_kb))
                pdf_path.unlink()
                return False

            return True

        except requests.RequestException as e:
            print("\n  [!] Intento {}/{} descarga fallida: {}".format(attempt, MAX_RETRIES, e))
            if attempt < MAX_RETRIES:
                time.sleep(2 ** attempt)

    return False


# --- Proceso principal --------------------------------------------------------

def main():
    parser = argparse.ArgumentParser(
        description="Descarga sentencias del TS (Sala Social) desde CENDOJ"
    )
    parser.add_argument(
        '--query', '-q', default=DEFAULT_QUERY,
        help='Termino de busqueda (default: "{}")'.format(DEFAULT_QUERY)
    )
    parser.add_argument(
        '--max-pages', '-p', type=int, default=MAX_PAGES,
        help='Maximo de paginas a recorrer (default: {})'.format(MAX_PAGES)
    )
    parser.add_argument(
        '--delay', '-d', type=float, default=DELAY_BETWEEN_REQUESTS,
        help='Segundos entre peticiones (default: {})'.format(DELAY_BETWEEN_REQUESTS)
    )
    parser.add_argument(
        '--skip-pdf', action='store_true',
        help='Solo extraer metadatos, no descargar PDFs'
    )
    parser.add_argument(
        '--sala-social', action='store_true',
        help='Filtrar solo resultados de la Sala de lo Social'
    )
    parser.add_argument(
        '--resume', action='store_true',
        help='Reanudar descarga desde el ultimo punto guardado'
    )

    args = parser.parse_args()
    query_encoded = quote(args.query, safe='')

    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
    PDF_DIR.mkdir(parents=True, exist_ok=True)

    progress = load_progress()
    existing_metadata = load_metadata()
    existing_roj = {s["roj"] for s in existing_metadata}

    print("=" * 70)
    print("DESCARGA DE SENTENCIAS - CENDOJ (Poder Judicial)")
    print("=" * 70)
    print('  Busqueda: "{}"'.format(args.query))
    print("  Paginas:  hasta {} (max. {} sentencias)".format(
        args.max_pages, args.max_pages * RESULTS_PER_PAGE))
    print("  Delay:    {}s entre peticiones".format(args.delay))
    print("  Salida:   {}".format(OUTPUT_DIR))
    if args.sala_social:
        print("  Filtro:   Solo Sala de lo Social")
    if args.skip_pdf:
        print("  PDFs:     NO (solo metadatos)")
    if args.resume and existing_metadata:
        print("  Reanudando: {} sentencias previas".format(len(existing_metadata)))
    print("=" * 70)
    print()

    session = requests.Session()
    all_sentencias = list(existing_metadata) if args.resume else []
    all_roj = set(existing_roj) if args.resume else set()
    new_count = 0
    skipped_count = 0

    # -- Fase 1: Scraping de metadatos -----------------------------------------

    print("FASE 1: Extraccion de metadatos de resultados de busqueda")
    print("-" * 50)

    for page in range(1, args.max_pages + 1):
        if args.resume and page in progress.get("pages_scraped", []):
            print("  Pagina {:2d}/{}: ya procesada, saltando".format(page, args.max_pages))
            continue

        offset = get_offset(page)
        url = SEARCH_URL.format(query=query_encoded, offset=offset)
        print("  Pagina {:2d}/{}: descargando...".format(page, args.max_pages))

        html = fetch_page(url, session)
        if html is None:
            print("  [X] No se pudo obtener la pagina {}".format(page))
            continue

        sentencias = parse_search_results(html, args.query)

        if not sentencias:
            print("  [X] Sin resultados en pagina {} -- fin de resultados".format(page))
            break

        if args.sala_social:
            before = len(sentencias)
            sentencias = [
                s for s in sentencias
                if 'social' in s.get('tipo_organo', '').lower()
            ]
            if before != len(sentencias):
                print("    (filtradas {} no-Social)".format(before - len(sentencias)))

        page_new = 0
        for s in sentencias:
            if s["roj"] not in all_roj:
                all_sentencias.append(s)
                all_roj.add(s["roj"])
                page_new += 1
            else:
                skipped_count += 1

        new_count += page_new
        print("    -> {} sentencias, {} nuevas".format(len(sentencias), page_new))

        progress.setdefault("pages_scraped", []).append(page)
        save_metadata(all_sentencias)
        save_progress(progress)

        if page < args.max_pages:
            time.sleep(args.delay)

    print()
    print("Total sentencias recopiladas: {} ({} nuevas)".format(len(all_sentencias), new_count))
    if skipped_count:
        print("  ({} duplicadas omitidas)".format(skipped_count))
    save_metadata(all_sentencias)

    if args.skip_pdf:
        print()
        print("[OK] Metadatos guardados en: {}".format(METADATA_FILE))
        return

    # -- Fase 2: Descarga de PDFs ----------------------------------------------

    print()
    print("FASE 2: Descarga de PDFs")
    print("-" * 50)

    to_download = [
        s for s in all_sentencias
        if s["roj"] not in progress.get("pdfs_downloaded", [])
        and s["roj"] not in progress.get("pdfs_failed", [])
    ]

    already = sum(1 for s in all_sentencias if (PDF_DIR / s["pdf_filename"]).exists())
    print("  {} ya descargados, {} pendientes".format(already, len(to_download)))
    print()

    success = already
    failed = 0

    for i, sentencia in enumerate(to_download, 1):
        roj = sentencia["roj"]
        pdf_path = PDF_DIR / sentencia["pdf_filename"]

        sys.stdout.write("  [{}/{}] {} ({})... ".format(
            i, len(to_download), roj, sentencia.get('fecha', '?')))
        sys.stdout.flush()

        if pdf_path.exists() and pdf_path.stat().st_size > 0:
            print("ya existe [OK]")
            progress.setdefault("pdfs_downloaded", []).append(roj)
            success += 1
            continue

        ok = download_pdf(sentencia, session)
        if ok:
            size_kb = pdf_path.stat().st_size / 1024
            print("[OK] ({:.0f} KB)".format(size_kb))
            progress.setdefault("pdfs_downloaded", []).append(roj)
            success += 1
        else:
            print("[FALLO]")
            progress.setdefault("pdfs_failed", []).append(roj)
            failed += 1

        save_progress(progress)
        time.sleep(args.delay)

    # -- Resumen final ---------------------------------------------------------

    print()
    print("=" * 70)
    print("RESUMEN")
    print("=" * 70)
    print("  Sentencias totales:   {}".format(len(all_sentencias)))
    print("  PDFs descargados:     {}".format(success))
    print("  PDFs fallidos:        {}".format(failed))
    print("  Metadatos:            {}".format(METADATA_FILE))
    print("  PDFs:                 {}".format(PDF_DIR))

    if failed > 0:
        print()
        print("  Para reintentar los fallidos, ejecuta de nuevo con --resume")
    print()


if __name__ == "__main__":
    main()
